### PR TITLE
Added SSEHandler to workaround UWSGI limitations

### PR DESCRIPTION
**Pull request by ehlertjd**
This commit spawns a thread for requests that need to send SSE events.
The thread then uses UWSGI offload threads to respond via Unix Socket.

See discussion [here](http://uwsgi-docs.readthedocs.io/en/latest/articles/OffloadingWebsocketsAndSSE.html) for why this is EVIL<sup>TM</sup>. 

One possible improvement would be to add a supervisor for SSE threads.

### Review Checklist

- Tests were added to cover all code changes
- Documentation was added / updated
- Code and tests follow standards in CONTRIBUTING.md

### CODE REVIEW

File: api/upload.py [hunk 1]
```diff
@@ -140,23 +141,9 @@ def process_upload(request, strategy, container_type=None, id_=None, origin=None
...
-        response.headers['Content-Type'] = 'text/event-stream; charset=utf-8'
-        response.headers['Connection']   = 'keep-alive'
-
-        # Instead of handing the iterator off to response.app_iter, send it ourselves.
-        # This prevents disconnections from leaving the API in a partially-complete state.
-        #
-        # Timing out between events or throwing an exception will result in undefinied behaviour.
-        # Right now, in our environment:
-        # - Timeouts may result in nginx-created 500 Bad Gateway HTML being added to the response.
-        # - Exceptions add some error json to the response, which is not SSE-sanitized.
```
Comment (kofalt) [inline api/upload.py]: Some of this doc block could be restored (the second paragraph).

Want to add a new first paragraph explaining the new setup?

File: api/web/sse_handler.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        try:
+            self.sock.bind(self.sock_file)
+            self.sock.listen(1)
+        except socket.error:
+            log.exception('Unable to open UNIX socket for SSE')
+            raise
+
+        response.headers['Content-Type'] = 'text/event-stream; charset=utf-8'
+        response.headers['Connection']   = 'keep-alive'
+        response.headers['X-SSE-Socket'] = self.sock_file
```
Comment (kofalt) [inline api/web/sse_handler.py]: Is this last header stripped by our nginx config? Does it make it that far?

Comment (ehlertjd) [reply api/web/sse_handler.py]: Header does not reach the client:

screen shot 2018-04-09 at 3 04 57 pm

File: api/upload.py (hunk 1 above)
Comment (ehlertjd) [reply api/upload.py]: @kofalt Good idea, updated.

File: api/web/sse_handler.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            self.client_sock.sendall(SSE_RESPONSE_HEADERS)
+        except: # pylint: disable=bare-except
+            # If we get connection errors, continue without sending events
+            log.exception('Error Initializing SSE response')
+            self.close()
+
+        try:
+            self.handler()
+        except: # pylint: disable=bare-except
+            log.exception('Error processing SSE response')
```
Comment (nagem) [inline api/web/sse_handler.py]: Will this prevent any named exceptions like `APINotFoundException` from hitting the `handle_exception` logic in the base handler? Although I'm not sure if PackfilePlacer has the potential to raise any.

**Comment by kofalt** [comment]
:+1: on latest commit

File: api/web/sse_handler.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+RECV_BUFFER_SIZE = 65535
+SSE_RESPONSE_HEADERS = 'HTTP/1.1 200 OK\r\nContent-Type: text/event-stream; charset=utf-8\r\nConnection: keep-alive\r\n\r\n'
+
+class SSEHandler(object):
+    def __init__(self, handler):
+        """ Perform the given handler function in another thread, responding with SSE events. 
+        NOTE: Any exceptions thrown by handler will be ignored, it is up to the handler to report any errors
+        using the send_event function. Perform an meaningful checks that should return a non-200 response BEFORE
+        invoking the handle function on this class.
```
Comment (nagem) [inline api/web/sse_handler.py]: 👍

**Comment by ehlertjd** [comment]
Closing in favor of #1133.

**Pull request closed without merging**